# Final Model

### Once I have coerced and generated features, and selected features, it is time to fit the model, and score the model against a holdout set.

### Pulling my data with selected scaled features

### Start by retrieving my data with selected scaled features in the previous step.

In [1]:
import pandas as pd

selected_scaled_df = pd.read_csv('https://raw.githubusercontent.com/mhan1/Data-Science/master/Machine%20Learning_Linear%20Regression%20project_Minyeong%20Han_Data%20Science/pruned_scaled_features.csv')
selected_scaled_df.head(3)

Unnamed: 0  symboling    height  curb_weight  engine_size      bore  \
0           0   1.743470 -2.020417    -0.014566     0.074449  0.519089   
1           1   1.743470 -2.020417    -0.014566     0.074449  0.519089   
2           2   0.133509 -0.543527     0.514882     0.604046 -2.404862   

   compression_ratio  horsepower  peak_rpm  highway_mpg  ...  make_nissan  \
0          -0.288349    0.171065 -0.263484    -0.546059  ...    -0.310253   
1          -0.288349    0.171065 -0.263484    -0.546059  ...    -0.310253   
2          -0.288349    1.261807 -0.263484    -0.691627  ...    -0.310253   

   make_peugot  make_volkswagen  make_volvo  body_style_other  \
0     -0.23812        -0.249351    -0.23812          3.693624   
1     -0.23812        -0.249351    -0.23812          3.693624   
2     -0.23812        -0.249351    -0.23812         -0.270737   

   drive_wheels_fwd  drive_wheels_rwd  engine_type_ohc  fuel_system_other  \
0         -1.188177          1.302831        -1.611363          -0.270737   
1         -1.188177          1.302831        -1.611363          -0.270737   
2         -1.188177          1.302831        -1.611363          -0.270737   

   normalized_losses  
0                0.0  
1                0.0  
2                0.0  

[3 rows x 23 columns]

In [2]:
#removing unnecessary column 
selected_scaled_df.drop('Unnamed: 0', axis=1, inplace=True)

In [3]:
selected_scaled_df.head().T

0         1         2         3         4
symboling                 1.743470  1.743470  0.133509  0.938490  0.938490
height                   -2.020417 -2.020417 -0.543527  0.235942  0.235942
curb_weight              -0.014566 -0.014566  0.514882 -0.420797  0.516807
engine_size               0.074449  0.074449  0.604046 -0.431076  0.218885
bore                      0.519089  0.519089 -2.404862 -0.517248 -0.517248
compression_ratio        -0.288349 -0.288349 -0.288349 -0.035973 -0.540725
horsepower                0.171065  0.171065  1.261807 -0.057230  0.272529
peak_rpm                 -0.263484 -0.263484 -0.263484  0.787346  0.787346
highway_mpg              -0.546059 -0.546059 -0.691627 -0.109354 -1.273900
engine_location_is_front  0.121867  0.121867  0.121867  0.121867  0.121867
fuel_type_is_gas          0.328798  0.328798  0.328798  0.328798  0.328798
make_honda               -0.260208 -0.260208 -0.260208 -0.260208 -0.260208
make_nissan              -0.310253 -0.310253 -0.310253 -0.310253 -0.310253
make_peugot              -0.238120 -0.238120 -0.238120 -0.238120 -0.238120
make_volkswagen          -0.249351 -0.249351 -0.249351 -0.249351 -0.249351
make_volvo               -0.238120 -0.238120 -0.238120 -0.238120 -0.238120
body_style_other          3.693624  3.693624 -0.270737 -0.270737 -0.270737
drive_wheels_fwd         -1.188177 -1.188177 -1.188177  0.841625 -1.188177
drive_wheels_rwd          1.302831  1.302831  1.302831 -0.767559 -0.767559
engine_type_ohc          -1.611363 -1.611363 -1.611363  0.620593  0.620593
fuel_system_other        -0.270737 -0.270737 -0.270737 -0.270737 -0.270737
normalized_losses         0.000000  0.000000  0.000000  1.328961  1.328961

In [4]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

selected_scaled_X = selected_scaled_df.iloc[:, :-1]
selected_scaled_y = selected_scaled_df.iloc[:, -1]

X_train, X_test, y_train, y_test = train_test_split(selected_scaled_X, selected_scaled_y, test_size=0.2, random_state=42)
X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.2, random_state=42)

model1 = LinearRegression()
model1.fit(X_train, y_train)
model1.coef_

array([ 0.43547727, -0.20557019, -0.10349567,  0.46819935, -0.11709273,
       -0.80084685, -0.3022446 ,  0.38447576, -0.09050541,  0.07469055,
       -0.81990147, -0.23078016,  0.13604241,  0.25502643, -0.09482945,
       -0.17682717, -0.28423826, -0.18667288,  0.41064767,  0.24811677,
       -0.25459872])

In [5]:
model1.score(X_test, y_test)

0.19558691730999467

In [6]:
model1.score(X_val, y_val)

0.28284018904824093

In [7]:
import numpy as np

def feature_importances(df, estimator, df_cols = None):
    df_cols = df_cols or df.columns
    df_cols = df_cols.to_numpy()
    coefs = estimator.coef_
    abs_coefs = abs(coefs)
    paired = np.hstack((df_cols.reshape(-1, 1), coefs.reshape(-1, 1), abs_coefs.reshape(-1, 1)))
    ordered_cols = paired[paired[:, -1].argsort()][::-1]
    return ordered_cols

In [8]:
feature_importances(selected_scaled_X, model1)

array([['fuel_type_is_gas', -0.8199014654813632, 0.8199014654813632],
       ['compression_ratio', -0.8008468518964104, 0.8008468518964104],
       ['engine_size', 0.46819934723664625, 0.46819934723664625],
       ['symboling', 0.43547727467117686, 0.43547727467117686],
       ['drive_wheels_rwd', 0.4106476746581554, 0.4106476746581554],
       ['peak_rpm', 0.38447575973113557, 0.38447575973113557],
       ['horsepower', -0.30224459773937806, 0.30224459773937806],
       ['body_style_other', -0.2842382643603246, 0.2842382643603246],
       ['make_peugot', 0.2550264323731658, 0.2550264323731658],
       ['fuel_system_other', -0.25459872016071416, 0.25459872016071416],
       ['engine_type_ohc', 0.24811676996959933, 0.24811676996959933],
       ['make_honda', -0.23078016197773016, 0.23078016197773016],
       ['height', -0.20557018933945712, 0.20557018933945712],
       ['drive_wheels_fwd', -0.18667288498171625, 0.18667288498171625],
       ['make_volvo', -0.17682717209517113, 0.17682717

### Inversing the scaled features

In [9]:
import pandas as pd

imputed_engineered_df = pd.read_csv('https://raw.githubusercontent.com/mhan1/Data-Science/master/Machine%20Learning_Linear%20Regression%20project_Minyeong%20Han_Data%20Science/imputed_engineered_features.csv')
imputed_engineered_df.head(3)

Unnamed: 0  symboling  num_of_doors  wheel-base  length  width  height  \
0           0          3           2.0        88.6   168.8   64.1    48.8   
1           1          3           2.0        88.6   168.8   64.1    48.8   
2           2          1           2.0        94.5   171.2   65.5    52.4   

   curb_weight  num_of_cylinders  engine_size  ...  engine_type_ohc  \
0         2548                 4          130  ...                0   
1         2548                 4          130  ...                0   
2         2823                 6          152  ...                0   

   engine_type_ohcf  engine_type_ohcv  engine_type_other  fuel_system_1bbl  \
0                 0                 0                  0                 0   
1                 0                 0                  0                 0   
2                 0                 1                  0                 0   

   fuel_system_2bbl  fuel_system_idi  fuel_system_mpfi  fuel_system_other  \
0                 0                0                 1                  0   
1                 0                0                 1                  0   
2                 0                0                 1                  0   

   normalized_losses  
0              122.0  
1              122.0  
2              122.0  

[3 rows x 50 columns]

In [10]:
unscaled_selected_df = imputed_engineered_df[selected_scaled_df.columns]
unscaled_selected_df.head(5).T

0        1        2        3        4
symboling                    3.00     3.00     1.00     2.00     2.00
height                      48.80    48.80    52.40    54.30    54.30
curb_weight               2548.00  2548.00  2823.00  2337.00  2824.00
engine_size                130.00   130.00   152.00   109.00   136.00
bore                         3.47     3.47     2.68     3.19     3.19
compression_ratio            9.00     9.00     9.00    10.00     8.00
horsepower                 111.00   111.00   154.00   102.00   115.00
peak_rpm                  5000.00  5000.00  5000.00  5500.00  5500.00
highway_mpg                 27.00    27.00    26.00    30.00    22.00
engine_location_is_front     1.00     1.00     1.00     1.00     1.00
fuel_type_is_gas             1.00     1.00     1.00     1.00     1.00
make_honda                   0.00     0.00     0.00     0.00     0.00
make_nissan                  0.00     0.00     0.00     0.00     0.00
make_peugot                  0.00     0.00     0.00     0.00     0.00
make_volkswagen              0.00     0.00     0.00     0.00     0.00
make_volvo                   0.00     0.00     0.00     0.00     0.00
body_style_other             1.00     1.00     0.00     0.00     0.00
drive_wheels_fwd             0.00     0.00     0.00     1.00     0.00
drive_wheels_rwd             1.00     1.00     1.00     0.00     0.00
engine_type_ohc              0.00     0.00     0.00     1.00     1.00
fuel_system_other            0.00     0.00     0.00     0.00     0.00
normalized_losses          122.00   122.00   122.00   164.00   164.00

In [11]:
#exporting the dataframe with unscaled_selected features as csv file for later use.
unscaled_selected_df.to_csv(r'C:\Users\Minyeong\Desktop\unscaled_selected_features.csv')

In [12]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

unscaled_selected_X = unscaled_selected_df.iloc[:, :-1]
unscaled_selected_y = unscaled_selected_df.iloc[:, -1]

unscaled_X_train, unscaled_X_test, unscaled_y_train, unscaled_y_test = train_test_split(unscaled_selected_X, unscaled_selected_y, test_size=0.2, random_state=42)
unscaled_X_train, unscaled_X_val, unscaled_y_train, unscaled_y_val = train_test_split(unscaled_X_train, unscaled_y_train, test_size=0.2, random_state=42)

model2 = LinearRegression()
model2.fit(unscaled_X_train, unscaled_y_train)
model2.coef_

array([ 1.10786826e+01, -2.66528024e+00, -6.29723772e-03,  3.56197596e-01,
       -1.36965178e+01, -6.38755741e+00, -2.42297690e-01,  2.55369257e-02,
       -4.16369194e-01,  1.96571402e+01, -8.73276569e+01, -2.99272761e+01,
        1.51917666e+01,  3.57667101e+01, -1.27662975e+01, -2.47994929e+01,
       -3.56117116e+01, -1.19749073e+01,  2.68694467e+01,  1.75016438e+01,
       -3.18982253e+01])

In [13]:
model2.score(unscaled_X_test, unscaled_y_test)

0.19558691730999464

In [14]:
model2.score(unscaled_X_val, unscaled_y_val)

0.2828401890482408

In [15]:
feature_importances(unscaled_selected_X, model2)

array([['fuel_type_is_gas', -87.3276568518106, 87.3276568518106],
       ['make_peugot', 35.76671011168316, 35.76671011168316],
       ['body_style_other', -35.61171162695837, 35.61171162695837],
       ['fuel_system_other', -31.89822532641943, 31.89822532641943],
       ['make_honda', -29.92727608851168, 29.92727608851168],
       ['drive_wheels_rwd', 26.869446700170244, 26.869446700170244],
       ['make_volvo', -24.799492920570554, 24.799492920570554],
       ['engine_location_is_front', 19.65714022643592, 19.65714022643592],
       ['engine_type_ohc', 17.501643776946057, 17.501643776946057],
       ['make_nissan', 15.191766616104044, 15.191766616104044],
       ['bore', -13.696517763111961, 13.696517763111961],
       ['make_volkswagen', -12.766297451916756, 12.766297451916756],
       ['drive_wheels_fwd', -11.97490730668438, 11.97490730668438],
       ['symboling', 11.078682583888966, 11.078682583888966],
       ['compression_ratio', -6.387557413665825, 6.387557413665825],
       

In [16]:
feature_importances(selected_scaled_X, model1)

array([['fuel_type_is_gas', -0.8199014654813632, 0.8199014654813632],
       ['compression_ratio', -0.8008468518964104, 0.8008468518964104],
       ['engine_size', 0.46819934723664625, 0.46819934723664625],
       ['symboling', 0.43547727467117686, 0.43547727467117686],
       ['drive_wheels_rwd', 0.4106476746581554, 0.4106476746581554],
       ['peak_rpm', 0.38447575973113557, 0.38447575973113557],
       ['horsepower', -0.30224459773937806, 0.30224459773937806],
       ['body_style_other', -0.2842382643603246, 0.2842382643603246],
       ['make_peugot', 0.2550264323731658, 0.2550264323731658],
       ['fuel_system_other', -0.25459872016071416, 0.25459872016071416],
       ['engine_type_ohc', 0.24811676996959933, 0.24811676996959933],
       ['make_honda', -0.23078016197773016, 0.23078016197773016],
       ['height', -0.20557018933945712, 0.20557018933945712],
       ['drive_wheels_fwd', -0.18667288498171625, 0.18667288498171625],
       ['make_volvo', -0.17682717209517113, 0.17682717

### model1 was trained with scaled_X and scaled_y after feature selection, and model2 was trained with unscaled_X and unscaled_y after the feature selection.<br>The score against the respective validation set was the same as 0.2828 for both model1 and model2. The score against the respective test set was also the same as 0.1956 for both model1 and model2.

### However, the coefficients for model1 and model2 were different, as shown above "feature_importance" output for each model.<br>For the model1 trained with scaled_X and scaled_y, 'fuel_type_is_gas' was the most important feature having the coefficient of -0.82 to predict the target variable "normalized_losses".<br> For the model2 trained with unscaled_X and unscaled_y, 'fuel_type_is_gas' is the most important feature but the coefficient was -87.33.<br> The other coefficients for each feature in respective model were different from each other, this is because the unit of each feature in the model2 is different from each other, while model1 has the same unit(z-score) for each feature. Hence, when it comes to talking about the feature importance, we should use scaled features like model1. When the unit is normalized, such as z-score, we can say how much our target variable change given an expected amount of movement in the feature.

### Through my analsis of the automobile data, I identified a set of core features that drive the normalized_loss values.<br>The target variable, normalized_loss represents the average loss per car per year. It is the relative average loss payment per insured vehicle year, and this value is normalized for all autos within a particular size classfication (two-door small, station wagons, sports/speciality, etc.)<br>According to the model1 feature importance,'fuel_type_is_gas', 'compression_ratio' were the top two features that change the normalized_loss value, followed by 'engine_size', 'symboling', 'drive_wheels_rwd', and so on.<br>The least important feature was 'engine_location_is_front', 'highway_mpg', 'make_volkswagen', 'curb_weight',  'bore', and so on.<br>I could see that the normalized losses in use as compared to other cars depend on symboling which was initially assigned risk factor to cars, with higher value if it is more risky. The negative cofficients for the top two important features such as 'fuel_type_is_gas' and 'compression_ratio' indicate that using gas as fuel type instead of diesel, and higher compression ratio (the ratio of the maximum to minimum volume in the cylinder of an internal combustion engine) contribute to lowering the normalized losses. A higher compression ratio means that the engine is getting more work out of the same amount of fuel, which is good for power and also miles per gallon. Higher compression ratio also means more thermal efficiency. Hence, it was good to confirm the negative coefficient for the 'compression_ratio' feature, as I could expect that higher thermal efficiency or miles per gallon could reduce the normalized losses.<br>In fact, 'horsepower' and 'highway_mpg' were also can be the measure for the thermal efficiency, and their respective coefficients appeared to be negative, as epxected. Going forward, I could reduce the features by eliminating these two 'horsepower' and 'highway_mpg' as 'compression_ratio' can represent these features and it is the factor that most contribute to normalized_loss among these three features.<br>Also, 'drive_wheels_rwd' and 'drive_wheels_fwd' appeared to drive normalized_loss, but with different sign. 'drive_wheels_rwd' increases the normalized_loss, while 'drive_wheels_fwd' decreases the normalized_loss based on the sign of respective coefficient.  A drive wheel is a wheel of a motor vehicle that transmits force, and front-wheel drive (FWD) vehicles' engines drive the front wheels. Using the front wheels for delivery of power as well as steering allows the driving force to act in the same direction as the wheel is pointing. Based on the different sign of coefficient for 'drive_wheels_rwd' and 'drive_wheels_fwd' features, I think this results can also help the car manufacturers to determine which type of drive wheel they should make.<br>Interestingly, several 'make' of cars appeared to contribute to the normalized_loss, with different coefficients, respectively. 'make_peugot' and 'make_nissan' have respective positive coefficients, which means these 'make' contribute to normalized_loss. Meanwhile, 'make_honda', 'make_volvo', 'make_volkswagen' have respective negative coefficients, meaning these 'make' contribute to lowering the normalized_loss. If I work for the insurance company, I would prefer 'honda', 'volvo', 'volkswagen' to other make, to reduce the normalized_loss value.<br>It was very interesting to see which feature drive the normalized_loss by using linear regression algorithm for this machine learning project.

#### References<br>
- https://jalopnik.com/heres-what-compression-ratio-actually-means-and-why-it-1819723873<br>
- https://en.m.wikipedia.org/wiki/Power_band<br>
- https://en.wikipedia.org/wiki/Drive_wheel#targetText=A%20drive%20wheel%20is%20a,vehicle%20moving%20forwards%20or%20backwards.